src/dbnode/storage/bootstrap/bootstrapper/fs/migrator/migrator.go

```diff
@@ -0,0 +1,261 @@
+		if err != nil {
+			return err
+		}
+		worker := &worker{
+			inputCh:        make(chan migrationCandidate, workerChannelSize),
+			outputCh:       outputCh,
+			persistManager: pm,
+			taskOptions: migration.NewTaskOptions().
+				SetFilesystemOptions(m.fsOpts).
+				SetStorageOptions(m.storageOpts),
+		}
+		workers = append(workers, worker)
+	}
+	closedWorkerChannels := false
+	closeWorkerChannels := func() {
+		if closedWorkerChannels {
+			return
+		}
+		closedWorkerChannels = true
+		for _, worker := range workers {
+			close(worker.inputCh)
+		}
+	}
+	// NB(nate): Ensure that channels always get closed.
+	defer closeWorkerChannels()
+
+	// Start up workers. Intentionally not using sync.WaitGroup so we can know when the last worker
+	// is finishing so that we can close the output channel.
+	activeWorkers := atomic.NewUint32(0)
+	for _, worker := range workers {
+		worker := worker
+		activeWorkers.Inc()
+		go func() {
+			m.startWorker(worker)
+			if activeWorkers.Dec() == 0 {
+				close(worker.outputCh)
+			}
+		}()
+	}
+
+	// Start up goroutine consuming worker output
+	var (
+		migrationResults = make(map[mergeKey]fs.ReadInfoFileResult, len(candidates))
+		resultsConsumed  = make(chan bool, 1)
+	)
+	go func() {
+		for result := range outputCh {
+			migrationResults[result.key] = result.updatedInfoFileResult
+		}
+		close(resultsConsumed)
+	}()
+
+	// Enqueue work for workers
+	for i, candidate := range candidates {
+		worker := workers[i%numWorkers]
+		worker.inputCh <- candidate
+	}
+
+	// Close channels now that work is enqueued
+	closeWorkerChannels()
+
+	// Wait until all workers have finished and migration results have been consumed
+	<-resultsConsumed
+
+	m.mergeUpdatedInfoFiles(migrationResults)
+
+	m.log.Info("fileset migration finished", zap.Duration("took", nowFn().Sub(begin)))
+
+	return nil
+}
+
+func (m *Migrator) findMigrationCandidates() []migrationCandidate {
+	var candidates []migrationCandidate
+	for md, resultsByShard := range m.infoFilesByNamespace {
+		for shard, results := range resultsByShard {
+			for _, info := range results {
+				newTaskFn, shouldMigrate := m.migrationTaskFn(info)
+				if shouldMigrate {
+					candidates = append(candidates, migrationCandidate{
+						newTaskFn:      newTaskFn,
+						metadata:       md,
+						shard:          shard,
+						infoFileResult: info,
+					})
+				}
+			}
+		}
+	}
+
+	return candidates
+}
+
+func (m *Migrator) startWorker(worker *worker) {
+	for input := range worker.inputCh {
+		task, err := input.newTaskFn(worker.taskOptions.
+			SetInfoFileResult(input.infoFileResult).
+			SetShard(input.shard).
+			SetNamespaceMetadata(input.metadata).
+			SetPersistManager(worker.persistManager))
+		if err != nil {
+			m.log.Error("error creating migration task", zap.Error(err))
+		}
+		infoFileResult, err := task.Run()
+		if err != nil {
+			m.log.Error("error running migration task", zap.Error(err))
+		}
+		worker.outputCh <- completedMigration{
```

Comment:
I'm a bit iffy about workers having a shared output channel in their structs here, could we have `outputCh chan <- completedMigration` as an arg to `startWorker` instead?

Reply:
Ah, that's a fair point. Yeah, I'll make that change.